docs/madr/decisions/031-locality-aware-load-balancing.md

```diff
@@ -0,0 +1,281 @@
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAware:
+      inZone: ["k8s.io/node", "k8s.io/az"]
+```
+
+#### Prefer traffic in single zone with fallback to other zones
+
+![Use case 3](assets/031/use_case_3.png)
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAware:
+        inZone: []
+    crossZone:
+      fallback: any
+```
+
+#### Load balance equally to list of zones
+
+![Use case 4](assets/031/use_case_4.png)
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAware:
+      inZone: []
+    crossZone:
+      - groups:
+        - ["zone-1", "zone-2"]
+```
+
+#### When leaving zone, target specific zones first
+
+![Use case 5](assets/031/use_case_5.png)
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAware:
+      inZone: []
+    crossZone:
+      - groups:
+        - ["zone-1"]
+      - groups:
+        - ["zone-3"]
+```
+
+#### Exclude single zone from load balancing
+
+![Use case 6](assets/031/use_case_6.png)
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAware:
+      inZone: []
+    crossZone:
+      - groups:
+        - ["zone-1"]
+      fallback: fail
+```
+
+TODO What if i have lots of zones and don’t want to write all of them during the incident. I would like to just exclude single zone, do we want to add something like:
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAware:
+      inZone: []
+    crossZone:
+      - exclude: ["zone-3"]
+```
+
+#### Load balance equally to local and zone
+
+![Use case 7](assets/031/use_case_7.png)
+
+We have the couple of use cases for this, but no config yet:
+- You are migrating from universal to k8s, both zones are in the same AZ and I would like to move traffic gradually from one to another
+- You have have some zone with testing or preview environment and you would like to handle live traffic on it for testing. Is it a use case for locality aware lb? Or should this be done with MeshHTTPRoute?
+The question is if we really need it, or can this be done currently in some other way?
+
+
+### Implementation
+
+We want to enhance current `MeshLoadBalancingStrategy` so code related to the changes should be placed in the plugin.
+
+#### Matching endpoints
+Each instance of the service might be in a different locality: zone, node. Users need to properly tag these dataplanes so we are able to match them. We need to retrieve all tags from all inbounds and, later, based on these tags, match outbounds that are in the same location.
+
+Example:
+
+On Kubernetes, all tags are populated automatically, while on Universal, users need to populate tags themselves:
```

Comment:
Have we checked this? I'm not sure labels are added to pods I think they are only available on the nodes. https://kubernetes.io/docs/concepts/services-networking/topology-aware-routing/

Reply:
I think I need to rephrase it, it's more about pods tag, thet are in the cluster. I think if a user doesn't have a node tag on the pod we might not get the value.